### PR TITLE
Dispose events

### PR DESCRIPTION
**Pull request by SvanT**
https://github.com/microsoft/vscode/issues/155232

The events that is setup for needs to be disposed for js to be able
to garbage collect the terminals.

### CODE REVIEW

**Comment by SvanT** [comment]
@Tyriar this combined with https://github.com/microsoft/vscode/pull/155976 seems to finally make it so that vscode garbage collects terminals directly after they are closed :)

File: addons/xterm-addon-webgl/src/renderLayer/CursorRenderLayer.ts [hunk 1]
```diff
@@ -54,6 +54,14 @@ export class CursorRenderLayer extends BaseRenderLayer {
     this.onOptionsChanged(terminal);
   }
 
+  public override dispose(): void {
+    if (this._cursorBlinkStateManager) {
+      this._cursorBlinkStateManager.dispose();
+      this._cursorBlinkStateManager = undefined;
+    }
```
Comment (Tyriar) [inline addons/xterm-addon-webgl/src/renderLayer/CursorRenderLayer.ts:61]: Little more concise:

```suggestion
    this._cursorBlinkStateManager?.dispose();
    this._cursorBlinkStateManager = undefined;
```

File: src/browser/Linkifier.ts [hunk 2]
```diff
@@ -53,6 +53,12 @@ export class Linkifier implements ILinkifier {
     };
   }
 
+  public dispose(): void {
+    this._onShowLinkUnderline.dispose();
+    this._onHideLinkUnderline.dispose();
+    this._onLinkTooltip.dispose();
+  }
```
Comment (Tyriar) [inline src/browser/Linkifier.ts:60]: The preferred way here is to have the class `extend Disposable` and register the emitters:

```ts
private _onShowLinkUnderline = this._register(new EventEmitter<ILinkifierEvent>());
```

File: src/browser/Types.d.ts [hunk 3]
```diff
@@ -201,6 +201,7 @@ export interface ILinkifier {
   linkifyRows(start: number, end: number): void;
   registerLinkMatcher(regex: RegExp, handler: LinkMatcherHandler, options?: ILinkMatcherOptions): number;
   deregisterLinkMatcher(matcherId: number): boolean;
+  dispose(): void;
```
Comment (Tyriar) [inline src/browser/Types.d.ts:204]: If we extend `Disposable` this interface could extend `IDisposable`

File: src/browser/Terminal.ts [hunk 4]
```diff
@@ -247,8 +247,12 @@ export class Terminal extends CoreTerminal implements ITerminal {
     super.dispose();
     this._renderService?.dispose();
     this._customKeyEventHandler = undefined;
+    this._overviewRulerRenderer?.dispose();
+    this._overviewRulerRenderer = undefined;
     this.write = () => { };
     this.element?.parentNode?.removeChild(this.element);
+    this.linkifier.dispose();
+    this.linkifier2.dispose();
```
Comment (Tyriar) [inline src/browser/Terminal.ts:255]: When extending Disposable:

```ts
    this.linkifier = this.register(this._instantiationService.createInstance(Linkifier));
    this.linkifier2 = this.register(this._instantiationService.createInstance(Linkifier2));
```

Looks like `linkifier2` is already getting registered.

File: src/browser/Terminal.ts [hunk 5]
```diff
@@ -247,8 +247,12 @@ export class Terminal extends CoreTerminal implements ITerminal {
     super.dispose();
     this._renderService?.dispose();
     this._customKeyEventHandler = undefined;
+    this._overviewRulerRenderer?.dispose();
+    this._overviewRulerRenderer = undefined;
```
Comment (Tyriar) [inline src/browser/Terminal.ts:251]: Best registered when it's created too

**Review by Tyriar — CHANGES_REQUESTED** [review]
Great finds, just some clean up/best practices

**Comment by SvanT** [comment]
@Tyriar Thanks for the feedback, I've updated the PR with the changes and I did a test that the disposal still works as expected.

Just a quick question, what is the difference of creating instances of a class through the InstantiationService instead of creating them using new Class() ?

**Comment by Tyriar** [comment]
@SvanT the instantiation service is used for dependency injection, when you call createInstance it will automatically fill in the params with the `@I...Service` decoration, like this:

https://github.com/xtermjs/xterm.js/blob/d2be5f992a639884f93520dd5f1ae2e9ae8c4730/src/browser/renderer/dom/DomRendererRowFactory.ts#L36-L44

**Comment by Tyriar** [comment]
I merged in the big v5 changes into master and this PR, `Linkifier` doesn't exist anymore so that's no longer needed here

**Review by Tyriar — APPROVED** [review]

**Pull request merged**